src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,510 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
-
-                // We only generate a switch dispatch if we have two or more value tests in a row
-                if (!(node is BoundTestDecisionDagNode firstTestNode &&
-                     firstTestNode.Test is BoundDagValueTest firstTest &&
-                     firstTestNode.WhenFalse is BoundTestDecisionDagNode whenFalse &&
-                     whenFalse.Test is BoundDagValueTest secondTest &&
-                     !loweredNodes.Contains(whenFalse) &&
-                     !this._dagNodeLabels.ContainsKey(whenFalse) &&
-                     firstTest.Input.Equals(secondTest.Input) &&
-                     firstTest.Input.Type.IsValidV6SwitchGoverningType()))
-                {
-                    // https://github.com/dotnet/roslyn/issues/12509 Could optimize float, double, decimal value switches. For now use if-then-else
+                if (!CanGenerateSwitchDispatch(node, loweredNodes))
                     return false;
+
+                var input = ((BoundTestDecisionDagNode)node).Test.Input;
+                ValueDispatchNode n = GatherValueDispatchNodes(node, loweredNodes, input);
+                LowerValueDispatchNode(n, input);
+                return true;
+            }
+
+            private bool CanGenerateSwitchDispatch(BoundDecisionDagNode node, HashSet<BoundDecisionDagNode> loweredNodes)
+            {
+                switch (node)
+                {
+                    // These are the forms worth optimizing.
+                    case BoundTestDecisionDagNode { WhenFalse: BoundTestDecisionDagNode test2 } test1:
+                        return CanDispatch(test1, test2);
+                    case BoundTestDecisionDagNode { WhenTrue: BoundTestDecisionDagNode test2 } test1:
+                        return CanDispatch(test1, test2);
+                    default:
+                        // Other cases are just as well done with a single test.
+                        return false;
+                }
+
+                bool CanDispatch(BoundTestDecisionDagNode test1, BoundTestDecisionDagNode test2)
+                {
+                    if (loweredNodes.Contains(test2))
+                    {
+                        Debug.Assert(this._dagNodeLabels.ContainsKey(test2));
+                        return false;
+                    }
+                    if (this._dagNodeLabels.ContainsKey(test2))
+                        return false;
+                    var t1 = test1.Test;
+                    var t2 = test2.Test;
+                    if (!(t1 is BoundDagValueTest || t1 is BoundDagRelationalTest))
+                        return false;
+                    if (!(t2 is BoundDagValueTest || t2 is BoundDagRelationalTest))
+                        return false;
+                    if (!t1.Input.Equals(t2.Input))
+                        return false;
+                    return true;
+                }
+            }
+
+            private ValueDispatchNode GatherValueDispatchNodes(
```

Comment:
>GatherValueDispatchNodes [](start = 38, length = 24)

It looks like this is only used in one location. Consider inlining.

Reply:
`fac` is an implementation detail I'd prefer to hide from the caller.

---
In reply to: [390492359](https://github.com/dotnet/roslyn/pull/42313#discussion_r390492359) [](ancestors = 390492359)